connector/servicegraphconnector/connector.go

```diff
@@ -0,0 +1,624 @@
+	}
+	defaultPeerAttributes = []string{
+		semconv.AttributeDBName, semconv.AttributeNetSockPeerAddr, semconv.AttributeNetPeerName, semconv.AttributeRPCService, semconv.AttributeNetSockPeerName, semconv.AttributeNetPeerName, semconv.AttributeHTTPURL, semconv.AttributeHTTPTarget,
+	}
+)
+
+type metricSeries struct {
+	dimensions  pcommon.Map
+	lastUpdated int64 // Used to remove stale series
+}
+
+var _ connector.Traces = (*serviceGraphConnector)(nil)
+
+type serviceGraphConnector struct {
+	config          *Config
+	logger          *zap.Logger
+	metricsConsumer consumer.Metrics
+
+	store *store.Store
+
+	startTime time.Time
+
+	seriesMutex                          sync.Mutex
+	reqTotal                             map[string]int64
+	reqFailedTotal                       map[string]int64
+	reqClientDurationSecondsCount        map[string]uint64
+	reqClientDurationSecondsSum          map[string]float64
+	reqClientDurationSecondsBucketCounts map[string][]uint64
+	reqServerDurationSecondsCount        map[string]uint64
+	reqServerDurationSecondsSum          map[string]float64
+	reqServerDurationSecondsBucketCounts map[string][]uint64
+	reqDurationBounds                    []float64
+
+	metricMutex sync.RWMutex
+	keyToMetric map[string]metricSeries
+
+	shutdownCh chan interface{}
+}
+
+func newConnector(logger *zap.Logger, config component.Config) *serviceGraphConnector {
+	pConfig := config.(*Config)
+
+	bounds := defaultLatencyHistogramBucketsMs
+	if pConfig.LatencyHistogramBuckets != nil {
+		bounds = mapDurationsToMillis(pConfig.LatencyHistogramBuckets)
+	}
+
+	if pConfig.CacheLoop <= 0 {
+		pConfig.CacheLoop = time.Minute
+	}
+
+	if pConfig.StoreExpirationLoop <= 0 {
+		pConfig.StoreExpirationLoop = 2 * time.Second
+	}
+
+	if pConfig.VirtualNodePeerAttributes == nil {
+		pConfig.VirtualNodePeerAttributes = defaultPeerAttributes
+	}
+
+	return &serviceGraphConnector{
+		config:                               pConfig,
+		logger:                               logger,
+		startTime:                            time.Now(),
+		reqTotal:                             make(map[string]int64),
+		reqFailedTotal:                       make(map[string]int64),
+		reqClientDurationSecondsCount:        make(map[string]uint64),
+		reqClientDurationSecondsSum:          make(map[string]float64),
+		reqClientDurationSecondsBucketCounts: make(map[string][]uint64),
+		reqServerDurationSecondsCount:        make(map[string]uint64),
+		reqServerDurationSecondsSum:          make(map[string]float64),
+		reqServerDurationSecondsBucketCounts: make(map[string][]uint64),
+		reqDurationBounds:                    bounds,
+		keyToMetric:                          make(map[string]metricSeries),
+		shutdownCh:                           make(chan interface{}),
+	}
+}
+
+func (p *serviceGraphConnector) Start(_ context.Context, host component.Host) error {
+	p.store = store.NewStore(p.config.Store.TTL, p.config.Store.MaxItems, p.onComplete, p.onExpire)
+
+	if p.metricsConsumer == nil {
```

Comment:
This looks odd: as a traces to metrics connector, I'd expect the metrics consumer to always be set.

Reply:
This should always be set -- this is also left over from the processor logic. I removed these nil checks in another place but missed this one.